src/kernel.rs

```diff
@@ -0,0 +1,100 @@
+//! The kernel type for filter operations
+
+/// An borrowed 2D kernel, used to filter images via convolution.
+#[derive(Debug, Copy, Clone)]
+pub struct Kernel<'a, K> {
+    pub(crate) data: &'a [K],
+    pub(crate) width: u32,
+    pub(crate) height: u32,
+}
+
+impl<'a, K> Kernel<'a, K> {
+    /// Construct a kernel from a slice and its dimensions. The input slice is
+    /// in row-major form.
+    pub fn new(data: &'a [K], width: u32, height: u32) -> Kernel<'a, K> {
+        assert!(width > 0 && height > 0, "width and height must be non-zero");
+        assert!(
+            width * height == data.len() as u32,
+            "Invalid kernel len: expecting {}, found {}",
+            width * height,
+            data.len()
+        );
+        Kernel {
+            data,
+            width,
+            height,
+        }
+    }
+
+    /// Get the value in the kernel at the given `x` and `y` position.
+    #[inline]
+    pub fn get_unchecked(&self, x: u32, y: u32) -> &K {
+        &self.data[(y * self.width + x) as usize]
+    }
+
+    /// The sobel horizontal 3x3 kernel.
+    pub const SOBEL_HORIZONTAL_3X3: Kernel<'static, i32> = Kernel {
```

Comment:
These should be top level constants. Being able to refer to a kernel of `i32` values via e.g. `Kernel::<u64>::SOBEL_HORIZONTAL_3x3` is very misleading.

Reply:
I'm a bit confused why the compiler even requires the disambiguation at all

Reply:
Actually what about implementing as associated constants of Kernel<i32> instead of impl<K> Kernel<K>?

Reply:
`kernel::Kernel::<i32>::SOBEL_HORIZONTAL_3x3` seems worse than `kernel::SOBEL_HORIZONTAL_3x3`. I don't see what the benefit is from nesting the consts.

Reply:
Nesting is necessary if for if we ever require kernels of different types other than `i32`. It's that or add types to the name of the constants such as `kernel::SOBEL_HORIZONTAL_3X3_i16` which seems worse.

Reply:
However, if you're confident that we will never need non-`i32` kernels or that this change should be postponed until we do need them then I'd be happy to make the constants module-level.

Reply:
Let's go with module level rather than trying to future proof this. Thanks.